### DISCUSSION TITLE
More transparancy around what the team internal roadmap

### DISCUSSION THREAD

**Discussion by gogainda** [oracle/truffleruby#2083] [General] [answered]
HI,
first of all thanks for the huge effort what you guys put into this project, this is really inspiring.

I would like to know better what is the current status of the project and were we are heading to. If you share the insights mb we can better organize and focus the community efforts on the achieving some particular results.

**Comment by claudiug** [comment]
@gogainda i dont think there is any public roadmap, and any clear direction. currently is: run ruby faster*


* terms and conditions.

**Comment by eregon** [comment, marked as answer]
Sorry for the late reply, I missed this post.
In general (and as always since the start) we're working on improving compatibility and performance.

Regarding compatibility, I think a wide number of gems now work out of the box or with small fixes (thanks for your issue reports and adding TruffleRuby in more gems' CIs!). I think it's becoming possible to run larger and larger applications on TruffleRuby with no changes to the Gemfile.
For C extensions, one limitation is related to mixing C extensions and precompiled native-only (no bitcode) static libraries. Currently this is not supported, but we have some ideas on how to solve it (compile more as bitcode, Sulong mixed mode, or turn the static library into a dynamic library which might need changes in the gem).

Recently we're more focused on performance and identified warmup as a major area where we need to improve.

Regarding Rack and Rails performance from [my recent talk](https://speakerdeck.com/eregon/running-rack-and-rails-faster-with-truffleruby), we aim to address:
* Faster warmup so an application gets to peak performance in a reasonable time. Various efforts here both in the Truffle framework and in TruffleRuby.
* Investigate and improve splitting (i.e., making a copy of a method's AST to specialize for a given caller), so that in general less splitting is needed since it hurts warmup if splitting too much, and that the default splitting limits are enough.
* Automatically not using a global lock for C extensions which are known to not need it. This needs some way to mark such C extensions.

With that work, we'll aim to run bigger applications and run them faster and smoothly on TruffleRuby.

**Reply by eregon** [reply]
I also updated the Aim and Status sections of the README to reflect this: https://github.com/oracle/truffleruby/commit/7fdfddeafd832587fb30a641a24dd1b776cfd368
